### ISSUE TITLE
Suggestion to add “-static” for windows (64-bit Pd is coming and all externals needs recompilation: they are all 32-bit)

### ISSUE THREAD

**Issue by Lucarda** [pure-data/pd-lib-builder#36]
As the windows 64-bit Pd is coming I suggest to include “-static” in the following lines:

`line 581`  c.ldflags = `-static` -static-libgcc -shared \

`line 589`  shared.ldflags = `-static` -static-libgcc -shared "$(pdbinpath)/pd.dll"

I have tested compilation for w32 with these lines added and externals worked without the need to ship dependencies.

As an example here is https://github.com/porres/pd-cyclone [coll]. Download pkg below:

[static-flag-tests.zip](https://github.com/pure-data/pd-lib-builder/files/1627558/static-flag-tests.zip)

The above pkg has different compilations. To test functionality close and reopen Pd on each test. (you can open ”coll-help.pd“ ::: make sure you don’t have another ”Cyclone“ somewhere in your path.)

[coll] is one case (as many others) that uses pthread and adding the “-static” flag nulls the need to ship pthread implementation.

Any thought against the “-static” flag?

**Comment by Lucarda** [comment]
>Also, can you explain why including the option in your own makefile(s) doesn't work?

Well you are right. We may close this issue.

**Comment by Lucarda** [comment]
I reopen the issue to report:

@umlaeute 

Unfortunately `-static` don’t work for w64.

https://stackoverflow.com/questions/48255993/undefined-reference-to-imp-pthread-create-when-cross-compiling-with-pthreads

https://sourceforge.net/p/mingw-w64/bugs/473/

**Comment by Lucarda** [comment]
Closing again.

**Comment by Lucarda** [comment]
## Good News for -static on w64


I was unable to do `-static` on w64 builds because for some reason I have an pthread.h on the pd/src. I think it shouldn't be there. I will tell @danomatika  this because he made the script that copies the sources. I deleted this file and succeed the compilation. With the file there there are incompatibilities because is different from the MinGW64's pthread.h.

Here's a depwalker of cyclone/coll getting rid of libwinpthread-1.dll on w64:

![staticoll](https://user-images.githubusercontent.com/25239522/35548802-a32855c6-0560-11e8-8510-10ebc9557ad8.jpg)



The reason I suggest `-static` is because if not, the [external] will use the pthread that comes with Pd. But is most likely that this will change in the future and the external will stop working.
The other alternative is to ship the [external]'s pthread, but this is the same as having it statistically linked and many developers may easily forget to ship it because they don't really know that is needed. For example now that Pd is using Msys2 and ships the pthread, anyone compiling an external with Msys2 will load the [external] with no trouble, not knowing that that particular pthread is needed and found in pd/bin. In the future Pd will use a new pthread and the [external] will fail to load.

There are many [objects] that use pthread.

this is line 633:   `c.ldflags := -static -static-libgcc -shared \`

**Comment by katjav** [comment]
While there's no 100% strict rule whether or not to define a setting in pdlibbuilder, this would be a simplified guideline:

1. A setting needed for all external libraries should be in Makefile.pdlibbuilder
2. A setting not needed for all libraries should go in the library makefile

In particular, an option that reduces the choice of options for others should go in the library makefile and not in Makefile.pdlibbuilder. Definition of `-static` in Makefile.pdlibbuilder would block the option for dynamic linking for all libs, and what works fine for one lib (or even most libs) could become a show stopper for someone else.

Therefore I ask you to try solve this in the library makefile and report your results. Some suggestions:

1. define `-static` in variable `ldflags` in library makefile
2. or not using `-static` at all, define `-lpthread` in variable `ldflags` rather than `ldlibs`
3. define `-lpthread` only for classes that need it, in variable `<classname>.class.ldflags`

Generalized: `ldflags` are set up for static linking and `ldlibs` for dynamic linking. Relevant comment section in Makefile.pdlibbuilder:

```
# cflags, ldflags, ldlibs:
# Define cflags (preprocessor&compiler), ldflags (linker) and ldlibs (dynamic
# link libs) for the whole library. These flags are added to platform-specific
# flags defined by Makefile.pdlibbuilder.
#
# <classname>.class.ldflags and <classname>.class.ldlibs:
# Define ldflags resp. ldlibs specific to class <classname>. These flags are
# added to platform-specific flags defined by Makefile.pdlibbuilder, and flags
# defined in your Makefile for the whole library. Note: cflags can not be
# defined per class in the current implementation.
```

**Comment by Lucarda** [comment]
@katjav 

>Therefore I ask you to try solve this in the library makefile and report your results

I succeeded tweaking this make file ---> https://github.com/porres/pd-cyclone/blob/master/Makefile line 10

to

uname := $(shell uname -s)
ifeq (MINGW,$(findstring MINGW,$(uname)))
ldlibs += -lpthread
`ldflags += -static`
exe.extension = .exe
endif

but:

>Definition of -static in Makefile.pdlibbuilder would block the option for dynamic linking for all libs, and what works fine for one lib (or even most libs) could become a show stopper for someone else.

I fail to imagine for which Lib this is a "show stopper". Thought I trust you, I think `-static` should be the default and a really minority of Libs, if any, should specify to cancel the` -static` flag.

## Some Knowledge Base:

Back in 2016 we had to update all w32 Deken packages to include the Pd-extended pthread implementation (along with other dlls.).
The thing was that Pd-extended installer copied many dlls to system folders. So people that didn't have Pd-extended installer were unable to load many [externals].
Here's the list of externals and deps that took me 3 weeks to detect.

[deken_w32_filled_2.txt](https://github.com/pure-data/pd-lib-builder/files/1683361/deken_w32_filled_2.txt)


So to not repeat the story in the future I believe is safe to include the `-static` flag as a default.
Also my point is that many developers will use pd-lib-builder for w32/64 and they will be coming mostly from Linux so they are mostly not aware of this Windows thing.

@umlaeute 

Is there really a problem making `-static` the default for pd-lib-builder?
The only effect that I see is that it auto-ships the pthread implementation.

**Comment by katjav** [comment]
> I fail to imagine for which Lib this is a "show stopper". Thought I trust you, I think `-static` should be the default and a really minority of Libs, if any, should specify to cancel the`-static` flag.

Dynamic linking is the compiler's default behavior which you can overrule with option `-static`. If  Makefile.pdlibbuilder defines `-static`, static linking becomes the default but it can not be overruled by a library makefile. The fact that we don't (currently) know who would prefer / need dynamic linking is not enough reason to take away freedom of choice.

**Comment by Lucarda** [comment]
> The fact that we don't (currently) know who would prefer / need dynamic linking is not enough reason to take away freedom of choice.

Ok, thanx for clarifying.

We might close this one and leave `-static` flag for the dev's makefiles.

**Comment by katjav** [comment]
> Is there really a problem making `-static` the default for pd-lib-builder?
> The only effect that I see is that it auto-ships the pthread implementation.

Maybe we misunderstand each other, and the documentation, which only says:

> On systems that support dynamic linking, this overrides  `-pie` and prevents linking
>  with the shared libraries. On other systems, this option has no effect. 

(https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html)

This reads as: "when option `-static` is given, no lib will be linked dynamically." And that is not what we want to force for everyone. In practice, when an archive is not found for static linking, the linker may resort to dynamic linking. So then it depends on your build setup what is linked statically / dynamically, which doesn't seem a good idea either.

I agree with you, it is confusing that `libpthread` is not included in standard libs for Windows (while it is included for other OS's), and to make things worse, that incompatible implementations of it exist. So, we should make it as convenient as can be within the given limitations. I could see the following approaches:

1. find the easiest syntax to include -static -lpthread from the lib makefile
2. figure out an equivalent of `-static-libgcc` for libptread, which doesn't affect other libs

For (1) I would suggest using the for<system> variable provided by pdlibbuilder, and try if `-static`  and `-lpthread` can both be defined in ldlibs:

```
define forWindows
  ldlibs := -static -lpthread
 <other Windows-specific defines>
endef
```
For (2) I do not know a method, internet may help us here.

**Comment by Lucarda** [comment]
Ok lets go slowly because I'm totally new to compiling.

1 (good) You seem to agree the need of statically link pthread (which has the same benefits as statically link G++ things as in #41)

2 (here we think diff) You seem to prefer, based on the freedom of choice, that `-static` should be an option on the makefile and not in plb. I understand your choice here because is difficult to supress the `-static` flag if it was placed on plb.

Now I think that if a library full of sources has no reference to pthread.h (ie none of the sources does threading) the `-static` flag has no effect.

And if another library has 7 sources with only one doing threading the `-static` flag will only affect the source with threading and not the others.

I ask:

Whats the difference on 
`ldlibs :=`
and
`ldflags :=`

My tests used `-static` on `ldflags`.

**Comment by Lucarda** [comment]
It will help a lot if we find a case when `-static` flag effectively ruins/mess an [external] or a lib.

I there any case which you suspect troubles?
We just test it.

**Comment by katjav** [comment]
> Now I think that if a library full of sources has no reference to pthread.h (ie none of the sources does threading) the `-static` flag has no effect.

That statement would be correct if libpthread is the only lib that you would ever want to link. But option `-static` applies to shared libs in general. E.g. libfftw3, libmp3lame, libogg would be linked statically if archives are found by the compiler. This should not be default behavior. It is bad enough that we have to link standard libs statically, making binaries up to ten times bigger for Windows than for other systems.

Statically linking standard libs is a clumsy workaround for the fact that Pd doesn't provide dlls for them. But since pthread is for other systems included in standard libs, it would make sense to treat it likewise (without affecting non-standard libs). If only there was an option `-static-libpthread` analoguous to `-static-libgcc`, we'd be done.

**Comment by Lucarda** [comment]
This needs further investigation.

I beleive that `-static` can only work with an **.a** file. There are some for that on Mingw32 for pthread.

pic showing *pthread search (this machine does not have MinGW64 so only showing the 32mingw)":
click to enlarge


![a1](https://user-images.githubusercontent.com/25239522/35710182-d12a5cd6-0794-11e8-977d-ce5bd8bef0f7.jpg)



I beleive `-static` can't work with an **dll**. If so Pierre could have make his "Camomile" for windows with a single dll and he couldn't. --> https://github.com/pierreguillot/Camomile/issues/65

**Comment by katjav** [comment]
> I beleive that `-static` can only work with an .a file.

Also called 'archive'. This is needed for static linking.

Here's an interesting discussion:

https://stackoverflow.com/questions/4156055/gcc-static-linking-only-some-libraries

Not sure which solution would apply if we want to link `libpthread` statically in pdlibbuilder, while not affecting other libs.

**Comment by Lucarda** [comment]
Yep, I beleive simply putting `-static` will not affect  " E.g. libfftw3, libmp3lame, libogg " because there are no .a for them and we are getting only what we want :: **static pthread**

correct?

Additional notes:
Msys2 compiled things are 10times bigger than the same ones compiled with Msys1.

**Comment by katjav** [comment]
Another thread:

https://stackoverflow.com/questions/809794/use-both-static-and-dynamically-linked-libraries-in-gcc

I guess that if we could define `-Wl -Bstatic -lpthread` or `-static -lpthread` at the end of the command line for the linker, it could work without affecting other libs. Not sure if the current order of variable expansion will allow this. It is a huge puzzle already, adding all flags from different origins.

**Comment by katjav** [comment]
> Additional notes:
> Msys2 compiled things are 10times bigger than the same ones compiled with Msys1.

That was my impression indeed. We might better campaign for standard libs dlls in pd/bin. For sure purr data won't like this bloat either.

**Comment by Lucarda** [comment]
> I guess that if we could define -Wl -Bstatic -lpthread or -static -lpthread at the end of the command line for the linker...

Looks good. I'll wait until you have somthing to test.

**Comment by Lucarda** [comment]
>We might better campaign for standard libs dlls in pd/bin

No, is better to not. it will be constatntly growing with older versions (also imposible as dome dlls has the same name)

**Comment by katjav** [comment]
I've opened isssue #42 which puts the current issue with `pthread` in the wider context of static standard libs, and a preliminary idea how to solve related issues.

**Comment by danomatika** [comment]
My 2 cents: definitely *do not* force static linking by default. That's against compiler conventions and should be left up to the external authors.

>Msys2 compiled things are 10times bigger than the same ones compiled with Msys1.

Are the resulting libs being stripped?

>We might better campaign for standard libs dlls in pd/bin

A pthread dll is included with the Pure Data vanilla build as it's needed for running Pd. I'd say it's safe to dynamically link to that unless there is a search path issue when the external is loaded. I'm not a big fan of forcing static link for a *very* common library that most *nix-style software needs to run on Windows, especially when we're already providing it with Pd. The new pthread dll is libwinpthread-1.dll and we should standardize on using that, as we do with libpd windows builds.

**Comment by Lucarda** [comment]
> definitely do not force static linking by default. That's against compiler conventions and should be left up to the external authors.

Externals authors will be mostly unaware that in the future when Msys2 libwinpthread has a new version and Pd adopts it, their external wont work anymore and needs to be recompiled with the new Msys2.

We already have this troubles back in 2016 (read above) and the solution was to ship an Msys1 pthreadGC-2.dll on every pkg that needs it. Statical linking might sound dirty but we must consider the benefits it has on the context of an Pd [external]. The other solution is to ship libwinpthread.dll

>Are the resulting libs being stripped?

Tested using the same makefile and source. I've searched and it looks that `--strip-unneeded` is OK for Msys2. I was expecting to find that Msys2 needed another flag.
